### ISSUE TITLE
Various minor issues with the build system

### ISSUE THREAD

**Issue by telamonian** [jupyterlab/jupyterlab#6698]
I've run into a bunch of issues over the last few months with the build system. I don't mind digging into it and trying to come up with fixes, but first I want to see if the issues I'm having are me problems (ie issues with my own config/env) or everyone problems.

General build system problems:

1. When opening sources in Chromium's devtools, I now only get compliled `.js` files instead of being able to see/debug the actual `.ts` files. I used to just get the `.ts` files without having to do anything special.

2. `jlpm run clean` doesn't fully clean up the existing build. There's been a number of times when I pulled updates from master and my build broke, and the only solution was to completely nuke my repo and start from a new clone; `jlpm run clean` was not sufficient.

Problems with the `--dev-mode --watch` build:

3. The `watch` build polls waaay too often, ends up consuming 100% of a cpu core at all times.

4. `watch` rebuilds files in response to changes, but somehow those rebuilt files aren't visible to my browser until I actually close and restart it; reloading the browser is not sufficient. This started happening sometime in the last few weeks, and is described more in #6649.

I already have fixes for `1)` and `3)`, but like I said, before I make a PR I want to see if anyone else is having these problems.

**Desktop:**

- OS: Mac 10.14.4
- Browser: Chromium 74
- JupyterLab: HEAD

**Comment by blink1073** [comment]
Thanks @telamonian!   I've seen 1 and 3 as well.  For 2, do you mean `jupyter lab clean`.  At one point we changed that to only remove staging.  I think we need an `--all` flag to remove static as well.

**Comment by blink1073** [comment]
Or did you mean this?

https://github.com/jupyterlab/jupyterlab/blob/c1593ac76c16fc9cb4490de1fcd5e380d6b4a4bf/package.json#L46

**Comment by blink1073** [comment]
We have a `clean:slate` that we use in the release process that is more agressive:

https://github.com/jupyterlab/jupyterlab/blob/c1593ac76c16fc9cb4490de1fcd5e380d6b4a4bf/package.json#L51

**Comment by telamonian** [comment]
I meant `jlpm run clean`, as in

https://github.com/jupyterlab/jupyterlab/blob/c1593ac76c16fc9cb4490de1fcd5e380d6b4a4bf/package.json#L46

but if that's not even really the right command (for prepping a guaranteed clean rebuild) then I guess that's `2)` "fixed" right there. I'll post a PR with the fixes for `1)` and `3)`, and see if I can also work out `4)`.

@blink1073 Relatedly, does `jlpm run clean:slate` respect "sibling" extension dev installs? Is there a way to pass it a general whitelist of things to leave alone?

**Comment by afshin** [comment]
I can corroborate the `--watch --dev-mode` reliability problems, it's been hitting me, too. But I don't know what the cause is.

**Comment by telamonian** [comment]
@afshin I've partially figured it out. If you add `--cache=false` to the `"watch"` command in `dev_mode/package.json`:

```
"watch": "webpack --watch --cache=false"
```

that "fixes" the `--watch --dev-mode` build. On the downside, it does push up the time for an incremental build to ~15 s. 

Basically, the problem seems to be that Webpack isn't able to figure out that the bundled files in `dev_mode/static` are ultimately dependent upon the `.ts` source files in the packages. Turning off Webpack's cache presumably just forces a complete rebuild of all of the bundled files (and probably some other things) at every incremental build.

I'm sure I can eventually figure out how to fix Webpack's dependency resolution for this case. For right now though, the `--cache=false` hack might be worthwhile, since the `--watch --dev-mode` build is useless in its current broken form.

@blink1073 Am I correct in thinking that you're the resident Webpack expert? Any ideas on how to fix this (apparent) dependency resolution issue in the `--watch` build?

**Comment by telamonian** [comment]
One more piece of information: I rolled back to 0f66b2860 (from March 1st), and the `--watch` build was miraculously restored. 

So the direct cause of the problem is either one of the changes made to the build system in the past three months (unfortunately, there are many), or a change in the Webpack version (0f66b2860 uses `4.12.0`, whereas HEAD uses `4.35.0`).

**Comment by blink1073** [comment]
The webpack version was my first guess as well. I'm fighting an illness, but I'll try and give full responses to your questions today.

**Comment by telamonian** [comment]
Getting closer. 4d1eebae7244674f0ee (June 1st) was no good, but the `--watch` build works in 463af7a3b8bde4cc3f2b (May 1st). It may be the Webpack version (for that commit it's `4.29.6`), but whatever is going on it appears that Webpack is failing to see most of the changes that occur following an incremental compilation.

To see what I mean, here's the output from a working incremental build (I made and saved a small tweak to `packages/codemirror/src/editor.ts`):

```
[3:05:58 PM] File change detected. Starting incremental compilation...

[3:06:10 PM] Found 0 errors. Watching for file changes.

Hash: 7a02065e326560edc651
Version: webpack 4.29.6
Child
    Hash: 7a02065e326560edc651
    Time: 1994ms
    Built at: 06/25/2019 3:06:11 PM
                               Asset      Size  Chunks             Chunk Names
                          index.html  1.54 KiB          [emitted]
        main.4ff22c98cf86fe6815e5.js  2.71 MiB    main  [emitted]  main
    main.4ff22c98cf86fe6815e5.js.map  2.74 MiB    main  [emitted]  main
     + 23 hidden assets
    Entrypoint main = vendors~main.bd76197dbd4292c367a0.js vendors~main.bd76197dbd4292c367a0.js.map main.4ff22c98cf86fe6815e5.js main.4ff22c98cf86fe6815e5.js.map
    [+06v] ../packages/codemirror/lib/factory.js 1.84 KiB {main} [built]
    [56HD] ../packages/codemirror/lib/codemirror-ipython.js 988 bytes {main} [built]
    [9FGT] ../packages/codemirror/lib/mimetype.js 1.23 KiB {main} [built]
    [P5RK] ../packages/codemirror/lib/index.js 557 bytes {main} [built]
    [ibHZ] ../packages/codemirror/lib/syntaxstatus.js 4.34 KiB {main} [built]
    [nzgn] ../packages/codemirror/lib/codemirror-ipythongfm.js 1.65 KiB {main} [built]
    [qb5G] ../packages/codemirror/lib/mode.js 4.58 KiB {main} [built]
    [yjD+] ../packages/codemirror/lib/editor.js 36.3 KiB {main} [built]
        + 2544 hidden modules

    WARNING in d3-array
      Multiple versions of d3-array found:
        1.2.4 /Users/tel/usr/venv/jup-old/git/jup-old/~/d3-geo/~/d3-array from /Users/tel/usr/venv/jup-old/git/jup-old/~/d3-geo/~/d3-array/src/index.js
        2.0.3 /Users/tel/usr/venv/jup-old/git/jup-old/~/d3-array from /Users/tel/usr/venv/jup-old/git/jup-old/~/d3-array/src/threshold/freedmanDiaconis.js

    Check how you can resolve duplicate packages:
    https://github.com/darrenscerri/duplicate-package-checker-webpack-plugin#resolving-duplicate-packages-in-your-bundle

    Child html-webpack-plugin for "index.html":
         1 asset
        Entrypoint undefined = index.html
           5 modules
Hash: 4398d95b85e2b87c9fdb
Version: webpack 4.29.6
Child
    Hash: 4398d95b85e2b87c9fdb
    Time: 1203ms
    Built at: 06/25/2019 3:06:13 PM
         Asset      Size  Chunks             Chunk Names
    index.html  1.54 KiB          [emitted]
     + 25 hidden assets
    Entrypoint main = vendors~main.bd76197dbd4292c367a0.js vendors~main.bd76197dbd4292c367a0.js.map main.4ff22c98cf86fe6815e5.js main.4ff22c98cf86fe6815e5.js.map
    [0igy] ../packages/documentsearch/lib/searchinstance.js 6.06 KiB {main} [built]
    [1vNz] ../packages/rendermime/lib/index.js 545 bytes {main} [built]
    [3hAR] ../packages/docregistry/lib/registry.js 22.8 KiB {main} [built]
    [4qND] ../packages/docregistry/lib/mimedocument.js 7.14 KiB {main} [built]
    [6ugk] ../packages/rendermime/lib/widgets.js 8.6 KiB {main} [built]
    [EKEa] ../packages/documentsearch/lib/searchproviderregistry.js 1.66 KiB {main} [built]
    [PoKp] ../packages/documentsearch/lib/providers/codemirrorsearchprovider.js 16.8 KiB {main} [built]
    [SDJn] ../packages/docregistry/lib/default.js 10.2 KiB {main} [built]
    [U+JU] ../packages/documentsearch/lib/providers/notebooksearchprovider.js 12.2 KiB {main} [built]
    [WyUK] ../packages/rendermime/lib/latex.js 6.66 KiB {main} [built]
    [YfE/] ../packages/docregistry/lib/context.js 23 KiB {main} [built]
    [e1iB] ../packages/rendermime/lib/mimemodel.js 1.17 KiB {main} [built]
    [hiU+] ../packages/codemirror-extension/lib/index.js 10.9 KiB {main} [optional] [built]
    [nPgm] ../packages/documentsearch/lib/index.js 314 bytes {main} [built]
    [rZA8] ../packages/docregistry/lib/index.js 246 bytes {main} [built]
        + 2537 hidden modules

    WARNING in d3-array
      Multiple versions of d3-array found:
        1.2.4 /Users/tel/usr/venv/jup-old/git/jup-old/~/d3-geo/~/d3-array from /Users/tel/usr/venv/jup-old/git/jup-old/~/d3-geo/~/d3-array/src/index.js
        2.0.3 /Users/tel/usr/venv/jup-old/git/jup-old/~/d3-array from /Users/tel/usr/venv/jup-old/git/jup-old/~/d3-array/src/threshold/freedmanDiaconis.js

    Check how you can resolve duplicate packages:
    https://github.com/darrenscerri/duplicate-package-checker-webpack-plugin#resolving-duplicate-packages-in-your-bundle

    Child html-webpack-plugin for "index.html":
         1 asset
        Entrypoint undefined = index.html
           5 modules
```

and here's the output from a non-working incremental build performed by HEAD:

```
[2:27:05 PM] File change detected. Starting incremental compilation...

[2:27:15 PM] Found 0 errors. Watching for file changes.

[0] Hash: de4f8d0b0fbfd08f2647                                                                                               
Version: webpack 4.35.0
Child
    Hash: de4f8d0b0fbfd08f2647
    Time: 1713ms
    Built at: 06/25/2019 2:27:17 PM
         Asset     Size  Chunks             Chunk Names
    index.html  1.6 KiB          [emitted]  
     + 25 hidden assets
    Entrypoint main = vendors~main.b85018d6e983eb251496.js vendors~main.b85018d6e983eb251496.js.map main.3d897e8e6dda1d48cb7e.js main.3d897e8e6dda1d48cb7e.js.map
    [hiU+] ../packages/codemirror-extension/lib/index.js 10.4 KiB {main} [optional] [built]
        + 1576 hidden modules
    Child html-webpack-plugin for "index.html":
         1 asset
        Entrypoint undefined = index.html
           5 modules
```

For the non-functional `--watch` build only a single file gets listed as rebuilt, and for whatever reason it doesn't trigger a rebuild of `{main}`, the chunk that contains it (possibly because the file is marked `[optional]`? Not sure).

Also, I can confirm that all of the same files in `packages/*/lib` are getting updated by the `--watch` build for both commits. Presumably the typescript compiler is picking up the changes, but Webpack isn't.

@blink1073 I'm trying to figure out what the tags (ie `{main} [optional] [built]`) in the Webpack output mean. `{main}` is obviously the name of the corresponding chunk, but the precise meaning of `[optional]` and `[built]` (built by the compiler? built into the chunk by Webpack?) escape me. Are you familiar with these annotations?

**Comment by telamonian** [comment]
Okay, I have the `--watch` build issue narrowed down, and it turns out it's something kind of dumb. It's not the Webpack version (regressing to an earlier version did nothing). It turns out the problem is with the `watchOptions.ignored` argument. Commenting it out completely actually fixes everything, no other intervention required.

Currently `watchOptions.ignored` is set up as a callback that takes a single argument (one path at a time) and returns `true` if something should be ignored and `false` otherwise. The problem is that as far as I can tell from [the Webpack docs](https://webpack.js.org/configuration/watch/#watchoptionsignored), callbacks are not supported. Instead, the only valid arguments for `watchOptions.ignored` are supposed to be `RegExp`, or an array of `anymatch` (fancy glob) strings. Possibly `ignore` can also take an array of `RegExp` or a single `anymatch`; the docs are frustratingly vague.

I tried digging through the Webpack source for more clues as to what the callback is actually doing. Though I did find out that watch is implemented via the [watchpack/watchpack](https://github.com/webpack/watchpack) package, I didn't find any code whatsoever that supports callbacks in `watchOptions.ignored`.

So I'm pretty confused. The simple solution would be to replace the callback with one or more regexes. However it would probably be impossible to completely reproduce the (intended, anyway) behavior of the callback that way. For right now I'll dig into the callback further and try to see what it's really doing and where it's going wrong.

@blink1073 It looks like you were the original author of the `ignore` callback, so if you could clue me in that would be nice. Also, sorry to bother you while you're sick.

**Comment by blink1073** [comment]
Hmm, I can't say for certain.  Perhaps because many other webpack parameters take functions I had assumed that one did as well.   We will need some kind of ignored pattern to avoid having a huge number of watched files.

I've never looked in to what those `[optional]` tags meant.

For `clean:slate`, I don't think we had considered sibling packages.   I like the idea of `clean` taking arguments about what to clean.

**Comment by telamonian** [comment]
I sprinkled some print statements into the `ignored` callback just to see what the heck was going on. The weirdest part is that somehow the callback actually was getting fed the file names, but each was getting repeated about 20 times. I'm also pretty sure that the return value was getting ignored.

My inclination was to nuke the callback and start over with some reasonable regexes (`/node_modules/` covers a whole lot of ground). However, It turns out that the current `ignored` callback is doing a whole bunch of stuff, at least some of which probably needs to be done somewhere. In particular, the last section is calling `maybeSync` on the files it deems appropriate:

https://github.com/jupyterlab/jupyterlab/blob/f0af9ca8a7371c48ff38a368942fe49c4d371a44/dev_mode/webpack.config.js#L205-L223

I guess this was just a convenient place to loop over the files in the build?

Not sure if it's wise, but I figured out a way to get a decent facsimile of the callback's expected filter behavior. I wrote a custom version of the built-in Webpack plugin `WatchIgnorePlugin` and spliced your `ignored` callback into it. This also has the advantage of leaving the `maybeSync` stuff undisturbed (if still maybe a little bit tangled).

I'm testing it out right now.

**Comment by blink1073** [comment]
Yeah, we can't just ignore node_modules like most folks because our linked extensions are in there

**Comment by telamonian** [comment]
> Yeah, we can't just ignore node_modules like most folks because our linked extensions are in there

Are you sure that's still the case during a `--dev-mode` and/or `--watch` build? I haven't dug into it deeply, but from what I've seen while debugging `webpack.config.js` in these cases the paths to the core packages are always of the form `${JLAB_REPO}/packages/@jupyterlab-x`.

 Granted, there are many fine details of the build's structure that still escape me.

**Comment by blink1073** [comment]
This webpack config is the same used by the built app, which doesn't have the yarn workspace structure

**Comment by telamonian** [comment]
Aha! Finally got it. Here's the problem lines:

https://github.com/jupyterlab/jupyterlab/blob/f0af9ca8a7371c48ff38a368942fe49c4d371a44/buildutils/src/ensure-repo.ts#L160-L163

These lines prevent any package without a `"jupyterlab"` entry in their `package.json` from being added to `linkedPackages`, which in turn thins out the `"linkedPackages"` object in `dev_mode/package.json`. `webpack.config.js` is set up to exclude from the `--watch` build any packages that aren't in `"linkedPackages"`. Since roughly 2/5 of the packages in core don't have a `"jupyterlab"` entry, `--watch` completely broke.

I pushed a fix over on #6700. I added a `"watchedPackages"` object to `dev_mode/package.json`, and `webpack.config.js` now uses that to figure out what files to watch. 

To start off I made the list in `"watchedPackages"` as expansive as was reasonable. Right now it's set up to watch all of the source files in all of the packages in the Jupyterlab repo, including stuff like `buildutils` and the packages in `tests`. In combination with setting the `watchOptions.poll` to poll only 3 times a second, the `--watch` build now only consumes about 6-8% of a CPU when idle. Which seems reasonable. I'll think more about whether the more expansive watch list is worth it.
